### PR TITLE
BUGFIX: SD-312 SD-347 disable stacking config with one measure

### PR DESCRIPTION
**Pull request by naibinh**
Changes:
- [SD-312](https://jira.intgdc.com/browse/SD-312), [SD-347](https://jira.intgdc.com/browse/SD-347): disable stacking config for dual axis chart with one measure on each axis
---
## PR Checklists
- [x] Squash commits
- [x] Successful unit test
- [x] Successful storybook test
- [x] Successful AD testcafe
- [x] Successful KD testcafe
---
## Related PRs
AD: https://github.com/gooddata/gdc-analytical-designer/pull/2125
KD: https://github.com/gooddata/gdc-dashboards/pull/1789

### CODE REVIEW

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3695/ : SUCCESS in 51s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2949/ : FAILURE in 23m 14s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1577/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3696/ : SUCCESS in 1m 00s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/976/ : UNSTABLE in 34m 31s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3709/ : SUCCESS in 31s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2962/ : FAILURE in 5m 57s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1581/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3710/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/980/ : UNSTABLE in 19m 57s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3711/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2964/ : SUCCESS in 3m 25s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1583/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3712/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/981/ : UNSTABLE in 16m 10s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3714/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2967/ : SUCCESS in 6m 10s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1585/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3715/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/982/ : SUCCESS in 16m 15s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3718/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2970/ : SUCCESS in 5m 26s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1587/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3719/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/983/ : UNSTABLE in 20m 43s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3721/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2972/ : SUCCESS in 4m 59s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1588/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3722/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/984/ : SUCCESS in 17m 34s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3723/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2973/ : SUCCESS in 2m 44s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1589/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3724/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/985/ : UNSTABLE in 13m 04s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3725/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2974/ : SUCCESS in 4m 01s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1590/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3726/ : SUCCESS in 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/986/ : SUCCESS in 15m 27s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3727/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2976/ : SUCCESS in 2m 25s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3740/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2990/ : SUCCESS in 6m 55s

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts [hunk 1]
```diff
@@ -4,91 +4,133 @@ import merge = require('lodash/merge');
...
 export const NORMAL_STACK = 'normal';
 export const PERCENT_STACK = 'percent';
 
-function handleStackMeasure(stackMeasures: boolean, item: any) {
+/**
+ * Set 'normal' stacking config to single series which will overwrite config in 'plotOptions.series'
+ * @param stackMeasures
+ * @param seriesItem
+ */
+function handleStackMeasure(stackMeasures: boolean, seriesItem: any) {
```
Comment (ptquang86) [inline src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: the type for seriesItem is ISeriesItem, right?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3749/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3000/ : SUCCESS in 7m 03s

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts [hunk 2]
```diff
@@ -4,91 +4,133 @@ import merge = require('lodash/merge');
...
     const hcStackingConfig = stacking ? { stacking: isFirstAxis ? stacking : NORMAL_STACK } : {};
 
     return {
-        ...item,
+        ...seriesItem,
         ...hcStackingConfig
     };
 }
 
+function countMeasuresInSeries(series: any): number[] {
```
Comment (ptquang86) [inline src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: is it `series: ISeriesItem[]`?

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts [hunk 3]
```diff
@@ -4,91 +4,133 @@ import merge = require('lodash/merge');
...
+        result[seriesItem.yAxis] += 1;
+        return result;
+    }, [0, 0]);
+}
+
+/**
+ * For y axis having one series, this series should be removed stacking config
+ * @param series
+ */
+function getSanitizedStackingSeries(series: any) {
```
Comment (ptquang86) [inline src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: same here

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts (hunk 2 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: Fixed

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts (hunk 3 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: Fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3757/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3006/ : SUCCESS in 16m 28s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3758/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3008/ : SUCCESS in 5m 13s

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3759/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3009/ : SUCCESS in 3m 47s

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts [hunk 4]
```diff
@@ -120,17 +120,17 @@ function getStackLabelPointsForDualAxis(stacks: any[]) {
...
         zip(...stacks.map((item: any, index: number) => values(item[`column${index}`])))
-    );
+    ).filter((point: any) => point);
 }
 
 function toggleStackedLabelsForDualAxis() {
     const { yAxis } = this;
 
-    const stackTotalGroups = yAxis.map((axis: any) => axis.stackTotalGroup);
+    const stackTotalGroups = yAxis.map((axis: any) => axis.stackTotalGroup).filter((group: any) => group);
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: move it to something like `getStackTotalGroups`

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts [hunk 5]
```diff
@@ -120,17 +120,17 @@ function getStackLabelPointsForDualAxis(stacks: any[]) {
...
     const stacks = yAxis.map((axis: any) => axis.stacks);
 
     if (stacks && stackTotalGroups) {
-        const points = getStackLabelPointsForDualAxis(stacks);
+        const points = getStackLabelPointsForDualAxis(stacks).filter((point: any) => point);
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: why not to put this filter directly inside of `getStackLabelPointsForDualAxis`?

File: src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts [hunk 6]
```diff
@@ -98,7 +98,42 @@ describe('getOptionalStackingConfiguration', () => {
...
+                    })
+                ],
+                yAxis: Array(2).fill({
+                    stackLabels: { enabled: true }
+                })
+            });
+        });
+
+        it('should return series without "stackMeasures" config with one measure in each axis for dual axis', () => {
             const chartOptions = {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts]: this `chartOptions` and `config` are used in two tests, why not to define them outside and share them. Try to keep tests as compact as possible

File: src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts [hunk 7]
```diff
@@ -118,20 +153,21 @@ describe('getOptionalStackingConfiguration', () => {
                 stackMeasuresToPercent: false,
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts]: apart of stackMeasuresToPercent this expect is same as in next test case, why not to share series and yAxis?

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts [hunk 8]
```diff
@@ -120,17 +120,17 @@ function getStackLabelPointsForDualAxis(stacks: any[]) {
     return flatten(
         // 'column0' is primary axis and 'column1' is secondary axis
         zip(...stacks.map((item: any, index: number) => values(item[`column${index}`])))
-    );
+    ).filter((point: any) => point);
 }
```
Comment (imnutz) [inline src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Use ```Lodash.identity``` function instead

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 4 above)
Comment (imnutz) [inline src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Use ```Lodash.identity``` function instead

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 5 above)
Comment (imnutz) [inline src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Use ```Lodash.identity``` function instead

File: src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts (hunk 6 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts]: I fixed by using `it.each()` for this case

File: src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts (hunk 7 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/test/getOptionalStackingConfiguration.spec.ts]: I fixed by using `it.each()` for this case

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 5 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Fixed

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 4 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Fixed

File: src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts [hunk 9]
```diff
@@ -4,90 +4,141 @@ import merge = require('lodash/merge');
...
+        result[seriesItem.yAxis] += 1;
+        return result;
+    }, [0, 0]);
+}
+
+/**
+ * For y axis having one series, this series should be removed stacking config
+ * @param series
+ */
+function getSanitizedStackingSeries(series: ISeriesItem[]): ISeriesItem[] {
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/getOptionalStackingConfiguration.ts]: Fixed

File: src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts [hunk 10]
```diff
@@ -24,8 +24,11 @@ export function isLabelOverlappingItsShape(point: any) {
 export const getDataLabelsGdcVisible = (chart: any): boolean | string =>
     get(chart, 'options.plotOptions.gdcOptions.dataLabels.visible', 'auto');
 
-export const areLabelsStacked = (chart: any) =>
-    (get(chart, 'userOptions.yAxis.0.stackLabels.enabled', false) && isStacked(chart));
+export const areLabelsStacked = (chart: any) => (
+    get(chart, 'userOptions.yAxis.0.stackLabels.enabled', false) ||
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/dataLabelsHelpers.ts]: Fixed

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 11]
```diff
@@ -1153,7 +1153,13 @@ export function getCategoriesForTwoAttributes(viewByTwoAttributes: IViewByTwoAtt
     }));
 }
 
-function getStackingConfig(stackByAttribute: any, options: IChartConfig): string {
+/**
+ * Get stacking config which will be set to 'highchart.plotOptions.series'
+ * in 'customConfiguration.ts:getStackingConfiguration()'
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: Fixed

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 5 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Fixed

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 4 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Fixed

File: src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts (hunk 8 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/autohideLabels/autohideColumnLabels.ts]: Fixed

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3763/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3013/ : SUCCESS in 8m 00s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/817/ : SUCCESS in 29s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3014/ : SUCCESS in 8m 06s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1606/ : SUCCESS in 22m 33s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/816/ : SUCCESS in 24m 47s

**Pull request merged**